apps/glusterfs/pendingop_entry.go

```diff
@@ -0,0 +1,265 @@
+		id = utils.GenUUID()
+	}
+	entry := &PendingOperationEntry{PendingOperation{
+		PendingItem: PendingItem{id},
+		Timestamp:   operationTimestamp(),
+		Actions:     []PendingOperationAction{},
+	}}
+	return entry
+}
+
+// NewPendingOperationEntryFromId fetches an existing pending operation entry
+// from the heketi db based on the provided id.
+func NewPendingOperationEntryFromId(tx *bolt.Tx, id string) (
+	*PendingOperationEntry, error) {
+	godbc.Require(tx != nil)
+	godbc.Require(id != "")
+
+	entry := &PendingOperationEntry{}
+	err := EntryLoad(tx, entry, id)
+	if err != nil {
+		return nil, err
+	}
+
+	if entry.Actions == nil {
+		entry.Actions = []PendingOperationAction{}
+	}
+
+	return entry, nil
+}
+
+// Save records the pending operation entry object in the db, keyed by the
+// value of its ID.
+func (p *PendingOperationEntry) Save(tx *bolt.Tx) error {
+	godbc.Require(tx != nil)
+	godbc.Require(p.Id != "")
+
+	return EntrySave(tx, p, p.Id)
+}
+
+// Delete removes a pending operation entry from the db.
+func (p *PendingOperationEntry) Delete(tx *bolt.Tx) error {
+	return EntryDelete(tx, p, p.Id)
+}
+
+// Marshal serializes the object for storage in the db.
+func (p *PendingOperationEntry) Marshal() ([]byte, error) {
+	var buffer bytes.Buffer
+	enc := gob.NewEncoder(&buffer)
+	err := enc.Encode(*p)
+
+	return buffer.Bytes(), err
+}
+
+// Unmarshal de-serializes the object from the db.
+func (p *PendingOperationEntry) Unmarshal(buffer []byte) error {
+	dec := gob.NewDecoder(bytes.NewReader(buffer))
+	err := dec.Decode(p)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// recordChange is a helper function to reduce some of the boilerplate around
+// adding a new change action item to the pending operation entry.
+func (p *PendingOperationEntry) recordChange(c PendingChangeType, id string) {
+	godbc.Require(p.Id != "")
+	godbc.Require(id != "")
+	p.Actions = append(p.Actions, PendingOperationAction{Change: c, Id: id})
+}
+
+// recordSizeChange is a helper function to reduce some of the boilerplate around
+// adding a new change action item that includes a size change to the entry.
+func (p *PendingOperationEntry) recordSizeChange(c PendingChangeType,
+	id string,
+	sizeDelta int) {
+
+	godbc.Require(p.Id != "")
+	godbc.Require(id != "")
+	p.Actions = append(p.Actions,
+		PendingOperationAction{
+			Change: c,
+			Id:     id,
+			Delta:  sizeDelta,
+		})
+}
+
+// RecordAddVolume adds tracking metadata for a new volume to the
+// PendingOperationEntry and VolumeEntry.
+func (p *PendingOperationEntry) RecordAddVolume(v *VolumeEntry) {
+	// track which volume this op is created
+	p.recordChange(OpAddVolume, v.Info.Id)
+	p.Type = OperationCreateVolume
+	// link back from "temporary" object to op
+	v.Pending.Id = p.Id
+}
+
+// FinalizeVolume removes tracking metadata from the volume entry.
+// This means that the volume is no longer pending.
+func (p *PendingOperationEntry) FinalizeVolume(v *VolumeEntry) {
+	v.Pending.Id = ""
```

Comment:
Spent some time thinking if it would help to have a DONE_ID like you have NEW_ID. Looks like there is not much use, as the objects start with a UUID. Hence, there are only two states "" denotes done, valid ID denotes in_progress. Right?

Reply:
Yes, in the case of NEW_ID I was trying to make explicit that passing an empty string into the New... function would make it allocate a new "guid".  In the case of setting x.Pending.Id = "" that's more standard (IMO) golang practice of empty string is the string type's zero value.

Reply:
+1 for the idea of integrating with rest op id, with your current model it can be achieved in upcoming releases.